File: guests/python/todos.py
"""The todos scene (tools/scenes/todos.steps). IT REGISTERS NO `on_undone`:
the derive's write rides the add's batch, so the ledger banks it."""

import sys
from dataclasses import dataclass

import kaya


@dataclass
class Todo:
    title: str
    done: bool


app = kaya.App()

draft = ""


def items_left_text(items):
    n = sum(1 for t in items.values() if not t.done)
    return "1 item left" if n == 1 else f"{n} items left"


def on_change(text):
    global draft
    draft = text


def on_add():
    if not draft:
        return
    # The marker leads the batch wherever the call sits in the body.
    kaya.undoable(f"add {draft}")
    # The binding mints the key (docs/fresh-key-plan.md).
    todos.insert_fresh(Todo(title=draft, done=False))
    # A SECOND transaction: `clear` inside an undoable group is refused at
    # apply, and a handler IS one transaction here.
    app.post(finish_form)


def finish_form():
    """The add's second transaction, on the app thread. CLEAR BEFORE
    FOCUS, so focus is the last word."""
    field.clear()
    field.focus()


def on_toggle(key, checked):
    # One field's delta: the title never travels.
    todos.patch(key, done=checked)


with app.window(title="todos"):
    with app.menu("Edit"):
        kaya.item("Undo", role=kaya.ROLE_UNDO)
        kaya.item("Redo", role=kaya.ROLE_REDO)

    todos = kaya.collection(Todo)
    items_left = todos.derive(items_left_text)

    with kaya.column():
        field = kaya.entry(on_change=on_change)
        kaya.button("Add", on_click=on_add)
        kaya.label(bind=items_left)
        # The body runs ONCE, authoring the blueprint.
        for todo in todos:
            with kaya.row():
                kaya.checkbox(checked=todo.done, on_toggle=on_toggle)
                kaya.label(bind=todo.title)

sys.exit(app.run())

File: guests/js/todos.ts
// The todos scene: IT REGISTERS NO `onUndone`, because the derive's write
// rides the add's batch.
//     KAYA_SELFTEST=todos node guests/js/todos.ts

import * as kaya from "kaya-gui";

const Todo = kaya.record({ title: String, done: Boolean }, "Todo");

const app = new kaya.App();

let draft = "";

function itemsLeftText(items: Map<kaya.Key, kaya.Fields<typeof Todo.schema>>): string {
  let n = 0;
  for (const t of items.values()) if (!t.done) n += 1;
  return n === 1 ? "1 item left" : `${n} items left`;
}

function onChange(text: string): void {
  draft = text;
}

async function onAdd(): Promise<void> {
  if (!draft) return;
  // The marker leads the batch wherever the call sits in the body.
  kaya.undoable(`add ${draft}`);
  // The binding mints the key (docs/fresh-key-plan.md).
  todos.insertFresh(Todo({ title: draft, done: false }));
  // A SECOND transaction: `clear` in an undoable group is refused, and the
  // handler's transaction commits at this await. CLEAR BEFORE FOCUS.
  await app.commit();
  field.clear();
  field.focus();
}

function onToggle(todo: kaya.RowHandle<kaya.Fields<typeof Todo.schema>>, checked: boolean): void {
  // One field's delta: the assignment IS the patch.
  todo.done = checked;
}

let field!: kaya.Widget;
let todos!: kaya.Collection<kaya.Fields<typeof Todo.schema>, kaya.Row<typeof Todo.schema>>;

app.window({ title: "todos" }, () => {
  app.menu("Edit", () => {
    kaya.item("Undo", { role: kaya.ROLE_UNDO });
    kaya.item("Redo", { role: kaya.ROLE_REDO });
  });

  todos = kaya.collection(Todo);
  const itemsLeft = todos.derive(itemsLeftText);

  kaya.column(() => {
    field = kaya.entry({ onChange });
    kaya.button("Add", { onClick: onAdd });
    kaya.label({ bind: itemsLeft });
    // The body runs ONCE, authoring the blueprint.
    for (const todo of todos) {
      kaya.row(() => {
        kaya.checkbox({ checked: todo.done, onToggle });
        kaya.label({ bind: todo.title });
      });
    }
  });
});

app.run();
